Feature automated body PR links

sushie1984 opened the pull request:
### Description

Addresses https://github.com/k3rnels-actions/pr-update/issues/165 by adding a `pr_body_with_links` input that allows to toggle whether the body of the PR from `prSource` to `prTarget` has the associated PRs links appended or not

### TODO

- [x] Fix issues with ci pipeline warnings
- [x] Create dist code
- [x] ENV for branch testing
- [x] Refactor to use https://docs.github.com/en/rest/commits/commits#compare-two-commits

sushie1984 committed:
add: prDiffUtils service class

This class performs the main logic of fetching all
associated PRs and using their meta data to retrieve
title and url.

sushie1984 committed:
rename: action and associated class

sushie1984 committed:
fix: issues regarding class naming

sushie1984 committed:
update: dist package code

sushie1984 committed:
fix: reported codefactor issues

sushie1984 committed:
add: dockerized setup for easier development

sushie1984 committed:
Merge pull request #2 from sushie1984/dockerize-setup

Add dockerized setup

sushie1984 committed:
add: ci test for link appending feature

sushie1984 committed:
fix: invalid inclusion of base PR in links

sushie1984 commented on .env.template:
```diff
@@ -0,0 +1,2 @@
+GITHUB_TOKEN=<create-your-own-token>
+GITHUB_REPOSITORY=sushie1984/pr-update-extended
```
Change to `k3rnels-actions/pr-update`

sushie1984 committed:
Merge branch 'k3rnels-actions:main' into main

sushie1984 committed:
Merge remote-tracking branch 'origin/main' into backup-feat-auto-pr-body-text

sushie1984 committed:
revert: GITHUB_REPOSITORY ENV to actual repo

sushie1984 committed:
refactor: using commit diff API

sushie1984 commented:
@k3rnelpan1c-dev 
1) Since I want to keep the tests as accurate as possible`__tests__/util/bodyUtils.test.ts` can you create a similar branch structure as I had in  https://github.com/sushie1984/pr-update/branches/all?query=test%2F ?

E.g.
* Having `prSource` (e.g. `test/staging`)
  * Having PR that is merged/squashed  `prSource`
* Having  `prTarget`  (e.g. `test/main`) 

Afterwards we could extract those parts (and the prNumber of the other tests) also into `env.ci/env.test` 

2) I have added also an [extended-test](https://github.com/k3rnels-actions/pr-update/pull/171/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R54) to the ci pipeline - let me know if this works out for you (or if there is a simpler way for a feature test)

3) Regarding your [comment](https://github.com/k3rnels-actions/pr-update/pull/166#pullrequestreview-1156426063) I added the missing Docker code - let me know if this ok for you (I am kind of a docker fanboy :smile:  which is why most of the projects that I touch have at least some docker context in it)

4) Regarding [your comment](https://github.com/k3rnels-actions/pr-update/pull/166#discussion_r1005685065) for the boolean input -> Still the same error (`TypeError`) when it is not present (although it is not required)

k3rnelpan1c-dev commented:
1. sure can/will do
2. oh? will take a closer look
3. I am a huge container fan myself (not of Docker.inc though), yet I sure will take another look :wink: 
4. sad, but it is what it is

k3rnelpan1c-dev commented:
1. sorry for the the delay, but I should have created the branches.
2. LGTM. To my knowledge there is no better way, as this is basically the only way to run such an integration for GHAs
3. Regarding the Docker related parts of this PR, while I get what you did there I still would like to ask that you exclude it from this PR. (if you are passionate about it we can discuss it in a followup)

sushie1984 committed:
remove: currently obsolete docker container feature

sushie1984 committed:
update: ci-workflow for easier test local setup

sushie1984 commented:
@k3rnelpan1c-dev 

> 1. sorry for the the delay, but I should have created the branches.

Thanks that should work - I have adjusted the test accordingly where necessary. 

I have also added an `.env.ci` that allows us to have the tests run against different repos (e.g. my forked version) when developing new stuff locally.

Also please have a look at  `.github/workflows/test.yml` as I updated the flow there so that `.env.ci` can be used -> I have added https://github.com/falti/dotenv-action to explicitly read the ENVs for the tests. Let me know if that works  (since I am not the repo owner you have to somehow whitelist this changes if you want to test beforehand)

> 2. LGTM. To my knowledge there is no better way, as this is basically the only way to run such an integration for GHAs

:+1: 

> 3. Regarding the Docker related parts of this PR, while I get what you did there I still would like to ask that you exclude it from this PR. (if you are passionate about it we can discuss it in a followup)

Sure, I have removed them and created [here](https://github.com/k3rnels-actions/pr-update/pull/175) an explicit draft PR that we can talk about this feature/development style.

k3rnelpan1c-dev commented on .github/workflows/test.yml:
```diff
@@ -20,6 +20,10 @@ jobs:
+      - name: Fetch ENV from .env.ci
+        uses: ./.github/actions/dotenv-action
+        with:
+          path: .env.ci
```
this would only work if we would include that action within this repo, however, the way easier approach would be the following:

```suggestion
      - name: Fetch ENV from .env.ci
        run: |-
          cat .env.ci >> $GITHUB_ENV
```

k3rnelpan1c-dev reviewed (changes requested):
besides the small test workflow comment, this LGTM :+1:

sushie1984 committed:
update: using simplified ENV read in ci tests

Co-authored-by: K3rnelPan1c <69395733+k3rnelpan1c-dev@users.noreply.github.com>

sushie1984 commented on .env.ci:
```diff
@@ -0,0 +1,3 @@
+GITHUB_REPOSITORY=k3rnels-actions/pr-update
+TEST_PR_UTIL_EXPECTED_PR_NUMBER=6
+TEST_BODY_UTILS_EXPECTED_PR_LINK=[add: "test-feature" to "staging"](https://github.com/k3rnels-actions/pr-update/pull/174)
```
@k3rnelpan1c-dev do you think we would need single quotes for this ENV? 
A quick test on my bash actually fails with a `syntax error near unexpected token `('`

```suggestion
TEST_BODY_UTILS_EXPECTED_PR_LINK='[add: "test-feature" to "staging"](https://github.com/k3rnels-actions/pr-update/pull/174)'
```

k3rnelpan1c-dev replied:
to be on the safe side, lets do that

sushie1984 committed:
add: single quotes on ENV containing special chars

k3rnelpan1c-dev commented:
Thank you very much for your contribution sushie!
